[Web Automation] Events may be dispatched to incorrect frame due to stale state

lilyspiniolas opened the pull request:
#### 76080cf19e9d456bf6eeb0d3b8b79cc4a3f67c7e
<pre>
[Web Automation] Events may be dispatched to incorrect frame due to stale state
<a href="https://bugs.webkit.org/show_bug.cgi?id=296037">https://bugs.webkit.org/show_bug.cgi?id=296037</a>
<a href="https://rdar.apple.com/155933183">rdar://155933183</a>

Reviewed by BJ Burg and Abrar Rahman Protyasha.

WebAutomationSessionProxy::focusFrame now takes an optional frameID instead of
a frameID, and focuses the main frame if frameID is not set. The method is
now called by WebAutomationSession::switchToBrowsingContext even when frameID
is empty so that focus is explicitly returned to the main frame.

At least one WPT test is known to have been affected by this bug when run via
webdriver: pointerevent_click_is_a_pointerevent.html?mouse has a failing subtest
as a result. With these changes, this test now completely passes.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::switchToBrowsingContext):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::focusFrame):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/297613@main">https://commits.webkit.org/297613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcc22779ce7ddc5ad04f15ca971b62de2e4301e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112136 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118163 "Built successfully") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62639 "An unexpected error occured. Recent messages:Printed configuration") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40388 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85193 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35934 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; layout-tests (failure); Running upload-test-results; layout-tests (failure); Compiled WebKit; layout-tests (failure); Running analyze-layout-tests-results") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25262 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62058 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95342 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121545 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39174 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29169 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94022 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39555 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93845 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24020 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39065 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16852 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35238 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39067 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44586 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38702 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42070 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40418 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -645,16 +645,30 @@ void WebAutomationSessionProxy::focusFrame(WebCore::PageIdentifier pageID, WebCo
-    // closing and it's not possible to focus the frame.
-    RefPtr frame = WebProcess::singleton().webFrame(frameID);
-    if (!frame || !frame->coreFrame() || !frame->coreFrame()->page()) {
+    auto executeHandlerWithFrameNotFoundError = [&] {
```
nit: just capture `completionHandler`

```suggestion
    auto executeHandlerWithFrameNotFoundError = [&completionHandler] {
```

aprotyas reviewed (approved):
Looks good to me, I suppose we should get coverage on the `test_no_browsing_context` failures in the WPT classic suite progressed by [296164@main](https://commits.webkit.org/296164@main) before landing.

aprotyas commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -477,14 +477,9 @@ void WebAutomationSession::switchToBrowsingContext(const Inspector::Protocol::Au
-        }
-
-        ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!WebFrameProxy::webFrame(frameID.value()), FrameNotFound);
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(frameID && !WebFrameProxy::webFrame(frameID.value()), FrameNotFound);
```
Hmmm.. how do we not fail this variant when `frameID == std::nullopt`?

aprotyas replied:
Oh, of course, `frameID && !WebFrameProxy::webFrame(frameID.value())` is the _error_ condition. Nvm

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -645,16 +645,30 @@ void WebAutomationSessionProxy::focusFrame(WebCore::PageIdentifier pageID, WebCo
+    RefPtr<WebCore::Frame> coreFrame;
+    if (frameID) {
+        RefPtr frame = WebProcess::singleton().webFrame(*frameID);
+        if (!frame) {
```
`ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!frame, FrameNotFound)`

lilyspiniolas replied:
ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF is used in WebAutomationSession, but I don't see it used anywhere in WebAutomationSessionProxy. Is it still desirable to update this? It looks like it would be distinct from the other methods if so.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -645,16 +645,30 @@ void WebAutomationSessionProxy::focusFrame(WebCore::PageIdentifier pageID, WebCo
+        coreFrame = page->mainFrame();
+
+    // If frame is no longer connected to the page, then it is
+    // closing and it's not possible to focus the frame.
```
`ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!coreFrame || !coreFrame->page(), FrameNotFound)`

lilyspiniolas committed:
[Web Automation] Events may be dispatched to incorrect frame due to stale state
https://bugs.webkit.org/show_bug.cgi?id=296037
rdar://155933183

Reviewed by BJ Burg and Abrar Rahman Protyasha.

WebAutomationSessionProxy::focusFrame now takes an optional frameID instead of
a frameID, and focuses the main frame if frameID is not set. The method is
now called by WebAutomationSession::switchToBrowsingContext even when frameID
is empty so that focus is explicitly returned to the main frame.

At least one WPT test is known to have been affected by this bug when run via
webdriver: pointerevent_click_is_a_pointerevent.html?mouse has a failing subtest
as a result. With these changes, this test now completely passes.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::switchToBrowsingContext):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::focusFrame):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:

Canonical link: https://commits.webkit.org/297613@main
